validate.go

```diff
@@ -54,9 +55,13 @@ func ValidateCid(c cid.Cid) error {
 		return ErrPossiblyInsecureHashFunction
 	}
 
-	if pref.MhType != mh.ID && pref.MhLength < minimumHashLength {
+	if pref.MhType != mh.IDENTITY && pref.MhLength < minimumHashLength {
 		return ErrBelowMinimumHashLength
 	}
 
+	if pref.MhType == mh.BLAKE3 && pref.MhLength != 32 {
```

Comment:
Blake3 is a variable sized hash function and AFAIK what we want is to be future proof in the decoding.

Right now it's architectureally complex to make the hash size of blake3 a parameter of `ipfs add` but hopefully what is easy is make `ipfs get` & friends correctly handle blake3 hashes of any sizes.

What I would do to do so is in `github.com/multiformats/go-multihash` I would make blake3's hasher produce hashes of the length you are comparing it to.

Reply:
I don't know how you want me to implement this in `go-multihash`- the multihash register thing requires a function with no arguments, that outputs a hasher. I think if people want to support other hash lengths they're going to need to export something that looks like `mh.BLAKE3_64` or something for each length they want. The use case I need only needs 32 bytes, which is standard.

Additionally, it doesn't really make sense to make blake3's hasher detect length and then produce hashes of the corresponding length because it just hashes one and then compares the bytes in all the multihash/set membership tests.

I looked into it and this is going to require a serious refactor of `go-multihash` and everything that uses its interface, and I'm really not trying to implement that right now...

Reply:
this is what it'd have to look vaguely like 

```// A wrapper is needed to get size variability
type blake3wrap struct {
	hash   hash.Hash64
	length int
}

func (x blake3wrap) Write(p []byte) (n int, err error) {
	return x.hash.Write(p)
}

func (x blake3wrap) Reset() {
	x.hash.Reset()
}

func (blake3wrap) BlockSize() int {
	return 1
}

func (x blake3wrap) Size() int {
	return int(x.length)
}

func (x blake3wrap) Sum(digest []byte) []byte {
	return x.hash.Sum(digest)
}

func init() {
	multihash.Register(multihash.BLAKE3, func(length int) hash.Hash { h := blake3.New(length, nil); return blake3wrap{h, length} })
}```

and then also where you hit the type errors with it

```Cannot use 'func(length int) hash.Hash { h := blake3.New(length, nil); return blake3wrap{h, length} }' (type func(length int) hash.Hash) as the type func() hash.Hash
```

Reply:
We can solve this one of two ways:

1. A refactor. But there's no reason this refactor would need to touch more than the multihash crate.
2. Make the blake3 hasher output "maximum size" digests, then truncate (which go-multihash will already do).

> What I would do to do so is in github.com/multiformats/go-multihash I would make blake3's hasher produce hashes of the length you are comparing it to.

As long as the default blake3 hasher produces, say, 64 or 128 byte hash digests by default, we don't need to worry. go-ipfs will call `Sum(data, code, length)` which will truncate to the desired length. If the requested digest size exceeds the maximum, we'll get a ErrLenTooLarge error.

For now, I'd be in favor of _not_ refactoring unless there's a large performance penalty to producing larger hashes.

Reply:
Regardless, we don't need _this_ check.

Reply:
Eh, ok. It didn't turn out too difficult: https://github.com/multiformats/go-multihash/pull/159